docs/advanced/topology.md

```diff
@@ -0,0 +1,125 @@
+
+One of the main goals of the topology service was to create an extensable caching mechanism,
+to store all aspects of infratructure with the ability for it to be easily accessed via API's.
+This allows implementors to write feature that are not bound to service providers API rate limits and latencies among other factors.
+Giving Clutch the control to provide a consistent user experience that you can control as your user base scales.
+
+## Topology Caching
+
+At the core of the topology service is its caching functionality which is the foundation that powers its API's.
+
+You can enable the cache like so, there are other [configuration](https://github.com/lyft/clutch/blob/c3097e5ad477952bb4bb90cc1fb5a126d7434565/api/config/service/topology/v1/topology.proto#L14-L28) options that can be tuned if necessary.
+
+```yaml title="clutch-config.yaml"
+services:
+  ...
+  # The topology services does require the postgres datastore to be configured
+  - name: clutch.service.topology
+    typed_config:
+      "@type": types.google.com/clutch.config.service.topology.v1.Config
+      // highlight-next-line
+      cache: {}
+```
+
+### Leader Election
+
+Currently Clutch elects a leader to handle the caching operations so that,
+only one instances performs write heavy operations.
+
+## How to extend the Topology Cache
+
+Extending the topology cache for private gateways can be done by satisfying the `CacheableTopology` [interface](https://github.com/lyft/clutch/blob/c3097e5ad477952bb4bb90cc1fb5a126d7434565/backend/service/topology/topology.go#L46-L57).
+
+```go
+type CacheableTopology interface {
+  CacheEnabled() bool
+  StartTopologyCaching(ctx context.Context, ttl time.Duration) (<-chan *topologyv1.UpdateCacheRequest, error)
+}
+```
+
+If satisfied, the topology service will start caching resources provided by the services satisfiying the interface and storing it in the `topology_cache` table to utilize for features like the [Search API](https://github.com/lyft/clutch/blob/c3097e5ad477952bb4bb90cc1fb5a126d7434565/api/topology/v1/topology_api.proto#L26-L32) which powers [autocomplete](#autocomplete).
+
+For completeness we can look at the Kubernetes service [caching implementation](https://github.com/lyft/clutch/blob/c3097e5ad477952bb4bb90cc1fb5a126d7434565/backend/service/k8s/cache.go#L28-L61) as an example of how this interface is satisfied.
+
+:::caution Config Load Order
+It's important to load the `clutch.service.topology` last or near the bottom in your Clutch configuration, this is because the topology service will iterate through the service registry to see if any satisfies the `CacheableTopology` interface to enable caching for those services.
+:::
+
+## Autocomplete
+
+Once the topology service and module are configured, autocomplete for all known Clutch resources and workflows will be enabled by default.
+
+```yaml title="clutch-config.yaml"
+modules:
+  // highlight-next-line
+ - name: clutch.module.topology
+services:
+  ...
```

Comment:
maybe throw in the k8s service as well without it highlighted to demonstrate ordering?